dartagnan/src/main/java/com/dat3m/dartagnan/verification/model/event/LocalModel.java

```diff
@@ -0,0 +1,13 @@
+package com.dat3m.dartagnan.verification.model.event;
+
+import com.dat3m.dartagnan.program.event.core.Local;
+import com.dat3m.dartagnan.verification.model.ThreadModel;
+
+import java.util.Set;
+
+
+public class LocalModel extends DefaultEventModel implements RegReaderModel, RegWriterModel {
+    public LocalModel(Local event, ThreadModel thread, int id) {
+        super(event, thread, id);
+    }
+}
```

Comment:
Shouldn't a `LocalModel` also capture the concrete value it has written to the register, just like `MemoryEventModel` captures its produced value?

Reply:
Where will be this feature used? In witness graph the default `String` of `Local` will be showed: `Register <- Expression`. Isn't is enough for witness?

Reply:
We have a transformation pass that flattens local computations.
If you have something like `r0 <- r1 >> r2 + zext(r3) ...` calculating the value manually would be painful. Thus I agree with Thomas that it would be good to use the value from the solver on the lhs. For the right hand side, I don't have a strong preference about keeping the register names (I have a slightly preference for this one) or their values

Reply:
> I'm not fully concerned about use-cases here, but just about completeness of the execution model: i.e., capturing all dynamic behaviour of the execution. 

I don't understand what you mean by this.

> That being said, I can see a use-case if we print nodes differently by overlaying dynamic information over the syntactic structure:
> 
> ```
> r2{8} <- r1{5} + r0{3}
> ```

I like this proposal, but we should probably use it everywhere where we have registers.

> 
> Another alternative is to write the syntax and the dynamic semantics separately like this:
> 
> ```
> r2 <- r1 + r0 // IR
> -------
> r2 <- 8        // Actual dynamic assignment
> ```
> 
> There are even more options to play around with.

Reply:
> > I'm not fully concerned about use-cases here, but just about completeness of the execution model: i.e., capturing all dynamic behaviour of the execution.
> 
> I don't understand what you mean by this.

The response was to this:
> Where will be this feature used?

For me the goal was never about visualizing witnesses or CAAT or any other use-case. I just want to have a class that captures a single execution of the program, i.e., all the values it produces/reads, all the communications that happened, the control-flow that was taken, the non-deterministic choices that were made, etc.
Once you have this, everything else will come into place naturally.

> 
> > That being said, I can see a use-case if we print nodes differently by overlaying dynamic information over the syntactic structure:
> > ```
> > r2{8} <- r1{5} + r0{3}
> > ```
> 
> I like this proposal, but we should probably use it everywhere where we have registers.

Of course. But as I said above, I'm not much concerned about how we eventually use this information: there are plenty of reasonable options.

Reply:
> For me the goal was never about visualizing witnesses or CAAT or any other use-case. I just want to have a class that captures a single execution of the program, i.e., all the values it produces/reads, all the communications that happened, the control-flow that was taken, the non-deterministic choices that were made, etc. Once you have this, everything else will come into place naturally.

I would say all those are covered by the current implementation:
- Locals (this include non-det choices), Loads and Stores are have the information about read/written values and their addresses
- The control flow is represented by the events in the model
- the communication that happened is covered by the relation model

Reply:
Yes, `Locals` were the only missing part as far as I see (one could be more explicit about non-det choices, but this is unnecessary for now). With them fixed, I think it looks good now.